### ISSUE TITLE
使用ray的grpo训练，没法用tensorboard，也没法存checkpoint

### ISSUE THREAD

**Issue by heningsu** [OpenRLHF/OpenRLHF#1006]
我使用slurm命令使用机器，脚本如下，但是正常运行之后没法用tensorboard，也没法存checkpoint。set -x

# reinforce++

ray job submit --address="http://octave:8265" \
   --runtime-env-json='{"working_dir": "/home/suliuhui/OpenRLHF"}' \
   -- python3 -m openrlhf.cli.train_grpo_ray \
   --ref_num_nodes 1 \
   --ref_num_gpus_per_node 1 \
   --reward_num_nodes 1 \
   --reward_num_gpus_per_node 1 \
   --actor_num_nodes 1 \
   --actor_num_gpus_per_node 1 \
   --vllm_num_engines 2 \
   --vllm_tensor_parallel_size 1 \
   --pretrain /home/suliuhui/Qwen2.5-3B-Instruct \
   --reward_pretrain OpenRLHF/Llama-3-8b-rm-700k \
   --save_path /openrlhf/examples/test_scripts/checkpoint/llama3-8b-rlhf \
   --micro_train_batch_size 16 \
   --train_batch_size 128 \
   --micro_rollout_batch_size 32 \
   --rollout_batch_size 1024 \
   --n_samples_per_prompt 5 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --max_samples 100000 \
   --generate_max_len 1024 \
   --init_kl_coef 1e-3 \
   --gamma 1.0 \
   --use_kl_loss \
   --kl_estimator k3 \
   --advantage_estimator group_norm \
   --zero_stage 3 \
   --bf16 \
   --actor_learning_rate 5e-7 \
   --prompt_data /home/suliuhui/MedicalGPT/medical-o1-reasoning-SFT/med_grpo/medical.jsonl \
   --input_key prompt \
   --label_key answer \
   --normalize_reward \
   --adam_offload \
   --gradient_checkpointing \
   --packing_samples \
   --save_steps 100 \
   --use_tensorboard /home/suliuhui/OpenRLHF/tensor-grpo-1 \
   --save_hf_ckpt \
   --ckpt_path /openrlhf/examples/test_scripts/ckpt/llama3-8b-rlhf

**Comment by hijkzzz** [comment]
save_steps  10

**Comment by heningsu** [comment]
我把reward函数改成了这样子def semantic_correctness(responses, answer, **kwargs):
    """Calculate semantic similarity using cross-encoder"""
    model = CrossEncoder('/home/suliuhui/MedicalGPT/medical-o1-reasoning-SFT/stsb-roberta-base', device=device)
    with torch.no_grad():   
        if responses is None:
            similarities= 0.0
        else:
            similarities = model.predict([responses,answer])
        return similarities
def extract_answer_from_model_output(text):
   parts = text.split("<answer>")#将文本按<answer>分割
   if len(parts) < 2:  # No <answer> tag found
       return None
   last_part = parts[-1] #提取最后一个<answer>标签后的内容
   if "</answer>" not in last_part:
       return None
   answer = last_part.split("</answer>")[0].strip() #提取<answer>和</answer>之间的内容
   return None if answer == "..." else answer   
#相似度计算
def correctness_reward(completions, answer, **kwargs): 
   #从 completions 中提取每个补全的文本内容
   responses = [completion[0]['content'] for completion in completions]
   extracted = [extract_answer_from_model_output(r) for r in responses]
   rewards = []
   for r, a in zip(extracted, answer):
       print("r:",r)
       print("a:",a)
       if r is None:
           simra = 0.0
       else:
           simra = semantic_correctness(r, a)*2.0
           
       rewards.append(simra)
    #    rewards.append(2.0)
   #对每个补全文本调用 .split() 方法，将其按空格分割为单词列表。计算单词列表的长度，表示补全的长度
   completion_lengths = [len(response.split()) for response in responses]
   return rewards

    
#据模型生成内容是否符合指定的XML格式分配奖励分数
def format_reward(completions, **kwargs):
   responses = [completion[0]['content'] for completion in completions]
   rewards = []
   format_scores = []
   for response in responses:
       score = 0.0
       if "<reasoning>" in response: score += 0.2
       if "</reasoning>" in response: score += 0.2
       if "<answer>" in response: score += 0.2
       if "</answer>" in response: score += 0.2
       rewards.append(score)
       format_scores.append(score)
   return rewards

#将正确性奖励和格式奖励结合，生成综合奖励分数
def combined_reward(prompts, completions, answer):
   # Get individual rewards
   correctness_scores = correctness_reward(completions=completions, answer=answer)
   format_scores = format_reward(completions=completions)
   # Combine rewards - correctness is weighted more heavily
   combined_rewards = []
   for c_score, f_score in zip(correctness_scores, format_scores):
       # Correctness score range: 0.0 to 2.0
       # Format score range: 0.0 to 0.8
       # Total range: 0.0 to 2.8
       combined_rewards.append(c_score + f_score)
   return combined_rewards

# # NOTE: reward function for multiple reward models, replace this with your own function!
# def reward_fn(rewards: List[torch.Tensor]):
#     return torch.stack(rewards).sum(dim=0)我就把bash脚本里面的reward_pretrain删了，但是ray启动服务之后，没使用gpu，好像跑不起来

**Comment by hijkzzz** [comment]
可以用 http remote rm

**Comment by heningsu** [comment]
修改之后还是卡在(LLMRayActor pid=3380115) INFO 05-06 15:26:59 [llm_engine.py:447] init engine (profile, create kv cache, warmup model) took 223.74 seconds 不开始训练，gpu使用也是0，但是前面加载时可以检测到cudaINFO 05-06 16:49:21 [__init__.py:256] Automatically detected platform cuda.
(ReferenceModelRayActor pid=3434920) You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with
 `model.to('cuda')`.

**Comment by CurryxIaoHu** [comment]
> 修改之后还是卡在(LLMRayActor pid=3380115) INFO 05-06 15:26:59 [llm_engine.py:447] init engine (profile, create kv cache, warmup model) took 223.74 seconds 不开始训练，gpu使用也是0，但是前面加载时可以检测到cudaINFO 05-06 16:49:21 [**init**.py:256] Automatically detected platform cuda. (ReferenceModelRayActor pid=3434920) You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.

请问解决了这个问题了吗？我也卡在这了
